src/worldfactory.cpp

```diff
@@ -1495,136 +1531,409 @@ int worldfactory::show_worldgen_tab_modselection( const catacurses::window &win,
+                                           ctxt.get_desc( "PICK_MODS", 1U ) ) );
+        wnoutrefresh( w_confirmation );
     } );
 
     do {
         ui_manager::redraw();
 
-        worldname = spopup.query_string( false );
-        const std::string action = ctxt.input_to_action( ctxt.get_raw_input() );
+        std::string action = ctxt.handle_input();
+        // Handle mouse input
         if( action == "MOUSE_MOVE" || action == "SELECT" ) {
-            sel_top_tab = 2;
-            cata::optional<point> coord = ctxt.get_coordinates_text( win );
+            cata::optional<point> coord = ctxt.get_coordinates_text( w_confirmation );
             if( !!coord ) {
-                int cnt = run_for_point_in<size_t, point>( tab_map, *coord,
-                [&sel_top_tab]( const std::pair<size_t, inclusive_rectangle<point>> &p ) {
-                    sel_top_tab = p.first;
+                run_for_point_in<int, point>( btn_map, *coord,
+                [&sel_opt]( const std::pair<int, inclusive_rectangle<point>> &p ) {
+                    sel_opt = p.first;
                 } );
-                if( cnt > 0 && action == "SELECT" && sel_top_tab != 2 ) {
-                    return sel_top_tab - 2;
+                if( action == "SELECT" ) {
+                    action = "CONFIRM";
+                    run_for_point_in<int, point>( slider_inc_map, *coord,
+                    [&action]( const std::pair<int, inclusive_rectangle<point>> &p ) {
+                        action = p.first % 2 == 0 ? "LEFT" : "RIGHT";
+                    } );
                 }
             }
-        } else if( action == "NEXT_TAB" ) {
-            if( worldname.empty() ) {
-                noname = true;
-                ui_manager::redraw();
-                if( !query_yn( _( "Are you SURE you're finished?  World name will be randomly generated." ) ) ) {
-                    noname = false;
-                    continue;
-                } else {
-                    noname = false;
-                    world->world_name = pick_random_name();
-                    if( !valid_worldname( world->world_name ) ) {
+        }
+
+        // Handle other inputs
+        if( action == "CONFIRM" ) {
+            if( sel_opt == 0 ) {
+                // rename
+                string_input_popup popup;
+                std::string message = popup
+                                      .max_length( max_worldname_len )
+                                      .title( _( "World Name:" ) )
+                                      .text( worldname )
+                                      .query_string();
+                if( !popup.canceled() ) {
+                    world->world_name = worldname = message;
+                }
+            } else if( sel_opt == static_cast<int>( sliders.size() + 1 ) ) {
+                // finish
+                if( worldname.empty() ) {
+                    noname = true;
+                    ui_manager::redraw();
```

Comment:
```suggestion
                    ui.invalidate_ui();
```

`query_yn` also calls `ui_manager::redraw()` so simply invalidating the UI should work.